rust/src/shards/gui/widgets/file_dialog.rs

```diff
@@ -0,0 +1,147 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2023 Fragcolor Pte. Ltd. */
+
+use super::FileDialog;
+use crate::core::cloneVar;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::BOOL_VAR_OR_NONE_SLICE;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::types::Context;
+use crate::types::ExposedTypes;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::NONE_TYPES;
+use crate::types::STRING_TYPES;
+
+lazy_static! {
+  static ref FILEDIALOG_PARAMETERS: Parameters = vec![
+    (
+      cstr!("Folder"),
+      shccstr!("To select a folder instead of a file."),
+      BOOL_VAR_OR_NONE_SLICE,
+    )
+      .into(),
+  ];
+}
+
+impl Default for FileDialog {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      folder: ParamVar::new(false.into()),
+    }
+  }
+}
+
+impl Shard for FileDialog {

```

Comment:
Why does this require a UI context, should be possible to use without I think

Reply:
Two reasons.

1. It is to make sure that it is only used in a context where we are able to draw UI (won't make sense for headless programs).
2. It isn't supported on all platforms (e.g. wasm) so we will need a egui replacement for those in the future.

Reply:
But I guess I can remove that constraint for now.

Reply:
I think makes sense for a headless program tbh, a shell program could use a picker, if UI is not needed don't need it.